exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+				errList = append(errList, err)
+				continue
+			}
+			for _, dataStr := range resourceSecret.Data {
+				dataList = append(dataList, dataStr)
+			}
+		default:
+			logger.Error(err, "Unsupported ClusterResourceSet resource kind", "Resource Kind", resource.Kind)
+			conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.UnsupportedResourceTypeReason, clusterv1.ConditionSeverityWarning, err.Error())
+			errList = append(errList, err)
+			continue
+		}
+
+		// Apply all values in the key-value pair of the resource to the cluster.
+		// As there can be multiple key-value pairs in a resource, each value may have multiple objects in it.
+		isSuccessful := true
+		for i := range dataList {
+			data := dataList[i]
+
+			if err := Apply(ctx, c, data); err != nil {
+				isSuccessful = false
+				logger.Error(err, "failed to apply ClusterResourceSet resource", "Resource kind", resource.Kind, "Resource name", resource.Name)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.ApplyFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+				errList = append(errList, err)
+			}
+		}
+
+		resourceInMap = *GenerateResourceStatus(isSuccessful, ComputeHash(dataList))
+		clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)] = resourceInMap
+	}
+	if len(errList) > 0 {
+		return kerrors.NewAggregate(errList)
+	}
+	return nil
+}
+
+// GenerateResourceStatus returns ResourceBinding to  be used in ClusterResourceSetBinding resource of the matching cluster
+func GenerateResourceStatus(isSuccessful bool, hash string) *addonsv1.ResourceBinding {
+	return &addonsv1.ResourceBinding{
+		Hash:            hash,
+		Applied:         isSuccessful,
+		LastAppliedTime: &metav1.Time{Time: time.Now().UTC()},
+	}
+}
+
+// GenerateResourceKindName generates a unique name to identify resources that is used in resources map in ClusterResourceSetBinding.
+func GenerateResourceKindName(resource addonsv1.ResourceRef) string {
+	return resource.Kind + "/" + resource.Name
+}
+
+// clusterToClusterResourceSet is mapper function that maps clusters to ClusterResourceSet
+func (r *ClusterResourceSetReconciler) clusterToClusterResourceSet(o handler.MapObject) []ctrl.Request {
+	result := []ctrl.Request{}
+
+	cluster, ok := o.Object.(*clusterv1.Cluster)
+	if !ok {
+		r.Log.Error(nil, fmt.Sprintf("Expected a Cluster but got a %T", o.Object))
+		return nil
+	}
+
+	resourceList := &addonsv1.ClusterResourceSetList{}
+	if err := r.Client.List(context.Background(), resourceList, client.InNamespace(cluster.Namespace)); err != nil {
+		r.Log.Error(err, "failed to list ClusterResourceSet")
+		return nil
+	}
+
+	clusterResourceSets := make([]*addonsv1.ClusterResourceSet, 0)
+	for i := range resourceList.Items {
+		m := &resourceList.Items[i]
+		if !HasMatchingLabel(cluster.GetLabels(), m.Spec.ClusterSelector.MatchLabels) {
```

Comment:
Similar comment as before - let's use all of ClusterSelector as there might be more than just MatchLabels.